### DISCUSSION TITLE
I use the stealth library, but how do I change the browser's default language?

### DISCUSSION THREAD

**Discussion by cplasfwst** [go-rod/rod#1057] [Q&A]
I use the stealth library, but how do I change the browser's default language?
The code is as follows. How should I modify it to change the browser's stealth to the language I want?

`package main

import (
	"github.com/go-rod/rod"
	"github.com/go-rod/rod/lib/devices"
	"github.com/go-rod/rod/lib/launcher"
	"github.com/go-rod/rod/lib/proto"
	"github.com/go-rod/stealth"
	"time"
)

var MyDevice = devices.Device{
	Title: "Chrome computer",
	//Capabilities:   []string{"touch", "mobile"},
	UserAgent: "Mozilla/5.0 (Windows NT 18.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
	//AcceptLanguage: "zh",
	Screen: devices.Screen{
		DevicePixelRatio: 2,
		Horizontal: devices.ScreenSize{
			Width:  1280,
			Height: 1024,
		},
		Vertical: devices.ScreenSize{
			Width:  1024,
			Height: 1280,
		},
	},
}

func main() {
	//path, _ := launcher.LookPath()
	//extPath, err := filepath.Abs("fixtures/chrome-extension")
	//if err != nil {
	//	fmt.Println("出错了:", err)
	//}
	u := launcher.New().Set("--disable-blink-features", "AutomationControlled").Set("no-first-run").Set("disable-default-apps").
		//Set("--user-agent", "Mozilla/5.0 (Windows NT 19.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36").
		//Set("--lang", "zh").
		//Env(append(os.Environ(), "lang=zh")...).
		//Set("load-extension", extPath).
		Headless(false).MustLaunch()
	//Headless(false).XVFB("--server-num=5", "--server-args=-screen 0 1600x900x16").MustLaunch()

	browser := rod.New().ControlURL(u).MustConnect()
	//targets, err := proto.TargetGetTargets{}.Call(browser)
	//if err != nil {
	//	log.Fatalln(err)
	//}
	//for _, t := range targets.TargetInfos {
	//	if t.Type != proto.TargetTargetInfoTypeBackgroundPage && t.Title != "Extension title" {
	//		return
	//	}
	//}
	// new page clear cookie，inject stealth.JS
	go browser.EachEvent(func(e *proto.TargetTargetCreated) {
		if e.TargetInfo.Type != proto.TargetTargetInfoTypePage {
			return
		}
		page := browser.MustSetCookies().MustPageFromTargetID(e.TargetInfo.TargetID)
		_, err := page.EvalOnNewDocument(TestJs)
		//_, err := page.EvalOnNewDocument(stealth.JS)
		if err != nil {
			return
		}
	})()

	browser.DefaultDevice(MyDevice)
	page := stealth.MustPage(browser)
	//page2 := stealth.MustPage(browser)
	//page3 := browser.MustPage(`https://bot.sannysoft.com/`)

	//stealth.MustPage(page)
	page.MustNavigate(`https://bot.sannysoft.com/`)
	//page2.MustNavigate(`https://bot.sannysoft.com/`)
	time.Sleep(time.Hour * 10)
	//fmt.Println(page3)
}
`

**Comment by ysmood** [comment]
https://github.com/go-rod/rod/blob/0340433eab4e6551f075eaf1875035469a777480/lib/devices/device.go#L13

**Comment by cplasfwst** [comment]
> https://github.com/go-rod/rod/blob/0340433eab4e6551f075eaf1875035469a777480/lib/devices/device.go#L13

After I used the stealth library, I used the AcceptLanguage string attribute again, but it did not take effect. The language used was also the stealth language.

**Reply by ysmood** [reply]
could you show us your code

**Comment by cplasfwst** [comment]
> could you show us your code

sure

code:

`package main

import (
	"github.com/go-rod/rod"
	"github.com/go-rod/rod/lib/devices"
	"github.com/go-rod/rod/lib/launcher"
	"github.com/go-rod/rod/lib/proto"
	"github.com/go-rod/stealth"
	"time"
)

var MyDevice = devices.Device{
	Title: "Chrome computer",
	//Capabilities:   []string{"touch", "mobile"},
	UserAgent: "Mozilla/5.0 (Windows NT 18.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36",
	//AcceptLanguage: "zh",
	Screen: devices.Screen{
		DevicePixelRatio: 2,
		Horizontal: devices.ScreenSize{
			Width:  1280,
			Height: 1024,
		},
		Vertical: devices.ScreenSize{
			Width:  1024,
			Height: 1280,
		},
	},
}

func main() {
	//path, _ := launcher.LookPath()
	//extPath, err := filepath.Abs("fixtures/chrome-extension")
	//if err != nil {
	//	fmt.Println("出错了:", err)
	//}
	u := launcher.New().Set("--disable-blink-features", "AutomationControlled").Set("no-first-run").Set("disable-default-apps").
		//Set("--user-agent", "Mozilla/5.0 (Windows NT 19.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36").
		//Set("--lang", "zh").
		//Env(append(os.Environ(), "lang=zh")...).
		//Set("load-extension", extPath).
		Headless(false).MustLaunch()
	//Headless(false).XVFB("--server-num=5", "--server-args=-screen 0 1600x900x16").MustLaunch()

	browser := rod.New().ControlURL(u).MustConnect()
	//targets, err := proto.TargetGetTargets{}.Call(browser)
	//if err != nil {
	//	log.Fatalln(err)
	//}
	//for _, t := range targets.TargetInfos {
	//	if t.Type != proto.TargetTargetInfoTypeBackgroundPage && t.Title != "Extension title" {
	//		return
	//	}
	//}
	// new page clear cookie，inject stealth.JS
	go browser.EachEvent(func(e *proto.TargetTargetCreated) {
		if e.TargetInfo.Type != proto.TargetTargetInfoTypePage {
			return
		}
		page := browser.MustSetCookies().MustPageFromTargetID(e.TargetInfo.TargetID)
		_, err := page.EvalOnNewDocument(stealth.JS)
		//_, err := page.EvalOnNewDocument(stealth.JS)
		if err != nil {
			return
		}
	})()

	browser.DefaultDevice(MyDevice)
	page := stealth.MustPage(browser)
	//page2 := stealth.MustPage(browser)
	//page3 := browser.MustPage(`https://bot.sannysoft.com/`)

	//stealth.MustPage(page)
	page.MustNavigate(`https://bot.sannysoft.com/`)
	//page2.MustNavigate(`https://bot.sannysoft.com/`)
	time.Sleep(time.Hour * 10)
	//fmt.Println(page3)
}
`

**Reply by ysmood** [reply]
could you use correct markdown format?

**Comment by cplasfwst** [comment]
Now my plan is to give up using stealth. I now hope to insert some commands in browser.EachEvent to bypass detection like https://bot.sannysoft.com/. Please tell me how I can detect Plugins Length through js code. (Old) 0
Plugins is of type PluginArray failed and is not 0?

**Comment by cplasfwst** [comment]
![image](https://github.com/go-rod/rod/assets/34654697/3c1cfb1c-459b-4964-b7c8-d0e449783e51)
